cache/cache.go

```diff
@@ -0,0 +1,159 @@
+package cache
+
+import (
+	"sync"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+var lock = &sync.Mutex{}
+var singleCache CacheBackend
+
+type CacheBackend interface {
+	SetCloudCache(string, CloudCache)
+	GetCloudCache(string) (CloudCache, bool)
+	SetServiceCache(string, string, ServiceCache)
+	GetServiceCache(string, string) (ServiceCache, bool)
+	SetMFCache(string, string, string, MFCache)
+	GetMFCache(string, string, string) (MFCache, bool)
+	FlushByCloud(time.Duration)
+	Clear()
+}
+
+// MetricFamily Cache Data
+type MFCache struct {
+	Time time.Time
+	MF   *dto.MetricFamily
+}
+
+// Service Cache Data
+type ServiceCache struct {
+	Time     time.Time
+	MFCaches map[string]*MFCache
+}
+
+// Cloud Cache Data
+type CloudCache struct {
+	Time          time.Time
+	ServiceCaches map[string]*ServiceCache
+}
```

Comment:
Is nested struct like this useful? Do you want to have 3 levels of cache: cloud level (cache all metrics within a cloud), service level (cache metrics per OS service), metrics level (cache each metric)?

I think caching whatever the `register.Gather` return is enough

Reply:
It's needed.

The exporter support multi clouds mode. You probably don't want to flush all the clouds in the same time?

Reply:
Yes, the original data `[]dto.MetricFamily`

Reply:
I think it can be like:

```go

type MetricFamilyCache struct {
    Cloud string
    Service string
    Time time.Time 
    Data *dto.MetricFamily
}

func (c *InMemoryCache) GetCache(cloud string, service string) (MetricFamilyCache, bool) {
    c.mu.Lock()
    defer c.mu.Unlock()
    if c.Cloud == cloud && c.Service == service {
        return *c.Data, true
    }
    return nil, false
}
```
what do you think?

Reply:
You miss the implementation details of `InMemoryCache`.
The `MetricFamilyCache` are only able to store one service's cache data. Where do you want to store all the caches?